test_agent/rust/src/main.rs

```diff
@@ -24,62 +24,62 @@
 mod testagent;
-use std::net::TcpStream as TcpStreamSync;
+use log::error;
+use std::net::TcpStream;
 use tokio::runtime::Runtime;
 
-fn main() {
-    let handle = thread::spawn(|| {
-        // Create a new Tokio runtime
-        let Ok(rt) = Runtime::new() else {
-            eprintln!("Error creating runtime");
-            return;
-        };
-
-        let test_agent_socket: TcpStreamSync =
-            TcpStreamSync::connect(TEST_MANAGER_ADDR).expect("issue in connecting  sync socket");
-        let test_agent_socket_to_tm: TcpStreamSync =
-            TcpStreamSync::connect(TEST_MANAGER_ADDR).expect("issue in connecting  sync socket");
+fn connect_to_socket(addr: &str, port: u16) -> Result<TcpStream, Box<dyn std::error::Error>> {
+    let socket_addr = format!("{addr}:{port}");
+    match TcpStream::connect(socket_addr) {
+        Ok(socket) => Ok(socket),
+        Err(err) => {
+            error!("Error connecting socket: {}", err);
+            Err(Box::new(err))
+        }
+    }
+}
 
-        rt.block_on(async {
-            // Spawn a Tokio task to connect to TEST_MANAGER_ADDR asynchronously
+async fn connect_and_receive() -> Result<(), Box<dyn std::error::Error>> {
+    let test_agent = connect_to_socket(TEST_MANAGER_ADDR.0, TEST_MANAGER_ADDR.1)?;
+    let ta_to_tm_socket = connect_to_socket(TEST_MANAGER_ADDR.0, TEST_MANAGER_ADDR.1)?;
+    let foo_listener_socket_to_tm = connect_to_socket(TEST_MANAGER_ADDR.0, TEST_MANAGER_ADDR.1)?;
 
-            let mut transport_socket = UTransportSocket::new();
-            let transport_socket_clone = transport_socket.clone();
+    let u_transport = UTransportSocket::new()?;
+    dbg!("Socket transport created successfully");
 
-            // Spawn a blocking task within the runtime
-            let blocking_task = tokio::task::spawn_blocking(move || {
-                println!("calling socket_init..");
-                transport_socket.socket_init();
-            });
+    let foo_listener = Arc::new(ListenerHandlers::new(foo_listener_socket_to_tm));
+    let agent = SocketTestAgent::new(test_agent, foo_listener);
+    agent
+        .clone()
+        .receive_from_tm(u_transport, ta_to_tm_socket)
+        .await;
 
-            // Don't wait for the blocking task to finish
-            tokio::spawn(async move {
-                if let Err(err) = blocking_task.await {
-                    dbg!("Error in socket_init: {}", err);
-                    return;
-                }
-                dbg!("socket_init completed successfully");
-            });
+    Ok(())
+}
 
-            let agent = SocketTestAgent::new(
-                test_agent_socket,
-                test_agent_socket_to_tm,
-                transport_socket_clone,
-            );
-            agent.clone().receive_from_tm().await;
-        });
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let handle = thread::spawn(|| {
+        let rt = Runtime::new().expect("Error creating runtime");
```

Comment:
This is an instance of where I think it's _not unreasonable_ to have an expect and panic.

If we're unable to create the Tokio Runtime, the program should probably exit.

There are ways to avoid the `.expect()` here, but they involve jumping through some hoops which we probably don't need to do here.

(If you're curious -- we'd have to create an async channel and communicate the result back to us on the main thread from the spawned thread. I've done it, but it's kinda clunky and a bit much for what we need here.)